### PR TITLE
impl(generator): add helper functions for formatting protobuf messages

### PR DESCRIPTION
**Pull request by scotthart**
part of the work for #10980

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11097)

### CODE REVIEW

File: generator/internal/discovery_type_vertex.cc [hunk 1]
```diff
@@ -48,6 +48,19 @@ void DiscoveryTypeVertex::AddNeededByTypeName(std::string type_name) {
   needed_by_.insert(std::move(type_name));
 }
 
+std::string DiscoveryTypeVertex::DetermineIntroducer(
+    nlohmann::json const& field) {
+  if (field.empty()) return "";
+  if (field.value("required", false)) return "";
+  if (field.contains("type") && field["type"] == "array") return "repeated ";
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.cc:55]: Seems like another good opportunity for `value()`.

File: generator/internal/discovery_type_vertex.cc [hunk 2]
```diff
@@ -48,6 +48,19 @@ void DiscoveryTypeVertex::AddNeededByTypeName(std::string type_name) {
...
+  // Test for map field.
+  if (field.contains("type") && field["type"] == "object" &&
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.cc:57]: Ditto.

File: generator/internal/discovery_type_vertex.h [hunk 3]
```diff
@@ -47,6 +47,10 @@ class DiscoveryTypeVertex {
   // field.
   void AddNeededByTypeName(std::string type_name);
 
+  // Returns "optional ", "repeated ", or and empty string depending on the
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.h:50]: s/and empty/an empty/

File: generator/internal/discovery_type_vertex.h [hunk 4]
```diff
@@ -47,6 +47,10 @@ class DiscoveryTypeVertex {
...
+  // field type.
+  static std::string DetermineIntroducer(nlohmann::json const& field);
```
Comment (devbww) [inline generator/internal/discovery_type_vertex.h:52]: Why should all these be `static` members, as opposed to non-members?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11097)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @devbww)

___
*[`generator/internal/discovery_type_vertex.h` line 50 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11097#-NRFXAcfb7xs-u1K8h-S-r1-50:-NRYoGu6EjqiHT_Bf_Ur:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1c008f3dc2ea7645dc76c099c23e5f3bc29ce6ff/generator/internal/discovery_type_vertex.h#L50)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

s/and empty/an empty/
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.h` line 52 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11097#-NRFXAcfb7xs-u1K8h-S-r1-52:-NRYoHFV6HRWXX0YLfxi:b-p5p7g8) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1c008f3dc2ea7645dc76c099c23e5f3bc29ce6ff/generator/internal/discovery_type_vertex.h#L52)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Why should all these be `static` members, as opposed to non-members?
</blockquote></details>

Originally I had them inside an anonymous namespace in the .cc file, but making them static member functions allowed testing to be more granular. Until it's proven these functions are needed elsewhere, I prefer having them exist in the scope in which they are used.
___
*[`generator/internal/discovery_type_vertex.cc` line 55 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11097#-NRFXAcfb7xs-u1K8h-R-r1-55:-NRYoHxG53cs_xf8039T:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1c008f3dc2ea7645dc76c099c23e5f3bc29ce6ff/generator/internal/discovery_type_vertex.cc#L55)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Seems like another good opportunity for `value()`.
</blockquote></details>

Done.
___
*[`generator/internal/discovery_type_vertex.cc` line 57 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11097#-NRFXAcfb7xs-u1K8h-R-r1-57:-NRYoIHq4YM7yEWaMwMH:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/1c008f3dc2ea7645dc76c099c23e5f3bc29ce6ff/generator/internal/discovery_type_vertex.cc#L57)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

Done.

**Review by devbww — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11097)* status: all files reviewed, 1 unresolved discussion (waiting on @devbww)

**Pull request merged**